diff --git a/hub/executor/executor.go b/hub/executor/executor.go
index b10b8b41..beaee6f8 100644
--- a/hub/executor/executor.go
+++ b/hub/executor/executor.go
@@ -106,7 +106,9 @@ func ApplyConfig(cfg *config.Config, force bool) {
 	updateNTP(cfg.NTP)
 	updateDNS(cfg.DNS, cfg.General.IPv6)
 	//updateListeners(cfg.General, cfg.Listeners, force)
-	//updateTun(cfg.General) // tun should not care "force"
+	if enableConcealRootTun() {
+		updateTun(cfg.General)
+	}
 	updateIPTables(cfg)
 	updateTunnels(cfg.Tunnels)
 
@@ -213,6 +215,11 @@ func updateTun(general *config.General) {
 	listener.ReCreateTun(general.Tun, tunnel.Tunnel)
 }
 
+func enableConcealRootTun() bool {
+	enabled, _ := strconv.ParseBool(os.Getenv("CONCEAL_FLCLASH_ROOT_TUN"))
+	return enabled
+}
+
 func updateExperimental(c *config.Experimental) {
 	if c.QUICGoDisableGSO {
 		_ = os.Setenv("QUIC_GO_DISABLE_GSO", strconv.FormatBool(true))
diff --git a/listener/sing_tun/server.go b/listener/sing_tun/server.go
--- a/listener/sing_tun/server.go
+++ b/listener/sing_tun/server.go
@@ -77,6 +77,12 @@ type ListenerHandler struct {
 	DisableICMPForwarding bool
 }
 
+type packageManagerCallback struct{}
+
+func (packageManagerCallback) OnPackagesUpdated(packages int, sharedUsers int) {
+	log.Debugln("[TUN] Android package manager updated: %d packages, %d shared users", packages, sharedUsers)
+}
+
 var emptyAddressSet = []*netipx.IPSet{{}}
 
 func CalculateInterfaceName(name string) (tunName string) {
@@ -409,6 +415,28 @@ func New(options LC.Tun, tunnel C.Tunnel, additions ...inbound.Addition) (l *Lis
 		EXP_SendMsgX:                          options.SendMsgX,
 	}
 
+	if runtime.GOOS == "android" && options.AutoRoute &&
+		(len(options.IncludeAndroidUser) > 0 || len(options.IncludePackage) > 0 || len(options.ExcludePackage) > 0) {
+		var packageManager tun.PackageManager
+		if len(options.IncludePackage) > 0 || len(options.ExcludePackage) > 0 {
+			packageManager, err = tun.NewPackageManager(tun.PackageManagerOptions{
+				Callback: packageManagerCallback{},
+				Logger:   log.SingLogger,
+			})
+			if err != nil {
+				err = E.Cause(err, "create package manager")
+				return
+			}
+			l.packageManager = packageManager
+			err = packageManager.Start()
+			if err != nil {
+				err = E.Cause(err, "start package manager")
+				return
+			}
+		}
+		tunOptions.BuildAndroidRules(packageManager, handler)
+	}
+
 	if options.AutoRedirect {
 		l.routeAddressMap = make(map[string]*netipx.IPSet)
 		l.routeExcludeAddressMap = make(map[string]*netipx.IPSet)
